src/main/java/bio/terra/workspace/db/ControlledResourceDao.java

```diff
@@ -0,0 +1,83 @@
+import bio.terra.workspace.service.controlledresource.model.ControlledResourceMetadata;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DuplicateKeyException;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ControlledResourceDao {
+  private final Logger logger = LoggerFactory.getLogger(ControlledResourceDao.class);
+  private final NamedParameterJdbcTemplate jdbcTemplate;
+
+  @Autowired
+  public ControlledResourceDao(NamedParameterJdbcTemplate jdbcTemplate) {
+    this.jdbcTemplate = jdbcTemplate;
+  }
+
+  public void createControlledResource(ControlledResourceMetadata controlledResource) {
+    final String sql =
+        "INSERT INTO workspace_resource "
+            + "(workspace_id, resource_id, associated_app, is_visible, owner, attributes) values "
+            + "(:workspaceId, :resourceId, :associatedApp, :isVisible, :owner, :attributes)";
+    final var params =
+        new MapSqlParameterSource()
+            .addValue("workspaceId", controlledResource.workspaceId())
+            .addValue("resourceId", controlledResource.resourceId())
+            .addValue("isVisible", controlledResource.isVisible())
+            .addValue("associatedApp", controlledResource.associatedApp().orElse(null))
+            .addValue("owner", controlledResource.owner().orElse(null))
+            .addValue("attributes", controlledResource.attributes().orElse(null));
+
+    try {
+      jdbcTemplate.update(sql, params);
+      logger.info(
+          String.format(
+              "Inserted record for workspaceResource %s",
+              controlledResource.resourceId().toString()));
+    } catch (DuplicateKeyException e) {
+      throw new DuplicateControlledResourceException(
+          String.format(
+              "Resource %s already exists in workspace %s",
+              controlledResource.resourceId(), controlledResource.workspaceId()),
+          e);
+    }
+  }
+
+  public Optional<ControlledResourceMetadata> getControlledResource(
+      UUID workspaceId, UUID resourceId) {
+    final String sql =
+        "SELECT workspace_id, resource_id, associated_app, is_visible, owner, attributes "
+            + "FROM workspace_resource "
+            + "WHERE workspace_id = :workspaceId AND resource_id = :resourceId";
+    final MapSqlParameterSource params =
+        new MapSqlParameterSource()
+            .addValue("workspaceId", workspaceId.toString())
+            .addValue("resourceId", resourceId.toString());
+    final Map<String, Object> columnToValue;
+    try {
+      columnToValue = jdbcTemplate.queryForMap(sql, params);
```

Comment:
`queryForMap` will throw if there aren't exactly 1 rows in the result set. We know there won't be more than one because the resource id is unique.

Reply:
Yes, but it leaves you casting later instead of being able to use the result methods like getString("workspace_id").

Reply:
True. So I'd be better off with a result mapper version like `public <T> T queryForObject(String sql, Map<String, ?> paramMap, RowMapper<T>rowMapper)`?